File: src/intelli/schemas/shell.py
"""Schemas for shell context and operator summary surfaces."""

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, Field


class ShellContextEntity(BaseModel):
    """Lightweight entity shown in breadcrumbs and switchers."""

    id: str
    label: str
    href: str | None = None
    subtitle: str | None = None


class ShellModelOption(BaseModel):
    """Selectable model option exposed to the shell."""

    id: str
    label: str
    provider: str | None = None
    is_default: bool = False


class ShellSessionSummary(BaseModel):
    """Recent or active session summary."""

    id: UUID
    module: str | None = None
    objective: str | None = None
    status: str
    last_active_at: datetime
    started_at: datetime


class ShellContextResponse(BaseModel):
    """Top-bar context for the product shell."""

    workspace_name: str
    workspace_id: str
    initiative_name: str | None = None
    active_project: ShellContextEntity | None = None
    active_task: ShellContextEntity | None = None
    active_session: ShellSessionSummary | None = None
    current_user_role: str | None = None
    available_models: list[ShellModelOption] = Field(default_factory=list)
    recent_entities: list[ShellContextEntity] = Field(default_factory=list)


class OpsActivityItem(BaseModel):
    """Recent operational activity for the shell overlay."""

    id: str
    label: str
    detail: str
    href: str | None = None
    timestamp: datetime
    kind: str = "run"
    status: str | None = None


class UsageSummaryItem(BaseModel):
    """Token and cost summary for a model or scope."""

    model: str
    input_tokens: int = 0
    output_tokens: int = 0
    cost_micros: int = 0


class OpsSummaryResponse(BaseModel):
    """Counts and recent activity for the operations overlay."""

    active_runs: list[OpsActivityItem] = Field(default_factory=list)
    recent_activity: list[OpsActivityItem] = Field(default_factory=list)
    queued_count: int = 0
    running_count: int = 0
    failed_count: int = 0
    unread_operational_items: int = 0
    price_table_version: str | None = None
    total_input_tokens: int = 0
    total_output_tokens: int = 0
    total_cost_micros: int = 0
    cost_by_model: list[UsageSummaryItem] = Field(default_factory=list)


class RightRailItem(BaseModel):
    """Single route-aware right-rail entry."""

    id: str
    title: str
    subtitle: str | None = None
    meta: str | None = None
    href: str | None = None
    badge: str | None = None


class RightRailSection(BaseModel):
    """Product-shell right-rail section."""

    kind: str
    title: str
    description: str | None = None
    items: list[RightRailItem] = Field(default_factory=list)


class RightRailResponse(BaseModel):
    """Route-aware right-rail payload."""

    module: str
    sections: list[RightRailSection] = Field(default_factory=list)

File: ui/src/types/shell.ts
export interface ShellContextEntity {
  id: string
  label: string
  href?: string | null
  subtitle?: string | null
}

export interface ShellModelOption {
  id: string
  label: string
  provider?: string | null
  is_default: boolean
}

export interface ShellSessionSummary {
  id: string
  module?: string | null
  objective?: string | null
  status: string
  last_active_at: string
  started_at: string
}

export interface ShellContext {
  workspace_name: string
  workspace_id: string
  initiative_name?: string | null
  active_project?: ShellContextEntity | null
  active_task?: ShellContextEntity | null
  active_session?: ShellSessionSummary | null
  current_user_role?: string | null
  available_models: ShellModelOption[]
  recent_entities: ShellContextEntity[]
}

export interface OpsActivityItem {
  id: string
  label: string
  detail: string
  href?: string | null
  timestamp: string
  kind: string
  status?: string | null
}

export interface UsageSummaryItem {
  model: string
  input_tokens: number
  output_tokens: number
  cost_micros: number
}

export interface OpsSummary {
  active_runs: OpsActivityItem[]
  recent_activity: OpsActivityItem[]
  queued_count: number
  running_count: number
  failed_count: number
  unread_operational_items: number
  price_table_version?: string | null
  total_input_tokens: number
  total_output_tokens: number
  total_cost_micros: number
  cost_by_model: UsageSummaryItem[]
}

export interface RightRailItem {
  id: string
  title: string
  subtitle?: string | null
  meta?: string | null
  href?: string | null
  badge?: string | null
}

export interface RightRailSection {
  kind: 'plans' | 'working_files' | 'sources' | 'activity' | 'usage' | 'empty_state' | string
  title: string
  description?: string | null
  items: RightRailItem[]
}

export interface RightRailPayload {
  module: string
  sections: RightRailSection[]
}
